### ISSUE TITLE
Maybe a bug in cache_position?

### ISSUE THREAD

**Issue by re-burn** [jungao1106/ICoT#28]
1. output in the previous step is not cached, causing `local variable 'outputs' referenced before assignment` in `image_attentions = torch.cat(self.attention_cache, dim=1).mean(dim=1)[:, -1]`
2. the size of image_attentions in the last dim is changing while generation, slice the image_attentions by `image_attentions = image_attentions[:, :self.query_image_mask.shape[1]]` before `image_attentions = image_attentions[self.query_image_mask]`
3. cache_position is updated only when ads is triggered, causing generation error after ads. Clone it and update it together with delta.
